### ISSUE TITLE
electrs on RaspiBlitz - running on the RaspberryPi 3

### ISSUE THREAD

**Issue by Himbeergeld** [openoms/bitcoin-tutorials#2]
Hello @openoms, just tried to install electrs on my RaspiBlitz V1.1 using your "electrs_install_on_RaspiBlitz.sh" script but I get stuck due to Connection refused errors:
*********************
Running `target/release/electrs -vvvv --index-batch-size=10 --jsonrpc-import --db-dir /mnt/hdd/electrs/db '--cookie=raspibolt:  <myPsssowordhere>' '--electrum-rpc-addr=0.0.0.0:50001'`
Config { log: StdErrLog { verbosity: Trace, quiet: false, timestamp: Off, modules: [], writer: "stderr", color_choice: Auto }, network_type: bitcoin, db_path: "/mnt/hdd/electrs/db/mainnet", daemon_dir: "/home/admin/.bitcoin", daemon_rpc_addr: V6([::1]:8332), cookie: Some("raspibolt:\u{1b}[C\u{1b}[C\u{1b}[C\u{1b}[D<myPasswordhere>"), electrum_rpc_addr: V4(0.0.0.0:50001), monitoring_addr: V6([::1]:4224), jsonrpc_import: true, index_batch_size: 10, bulk_index_threads: 4, tx_cache_size: 10000, txid_limit: 100, server_banner: "Welcome to electrs (Electrum Rust Server)!" }
DEBUG - Server listening on [::1]:4224
DEBUG - Running accept thread
WARN - failed to connect daemon at [::1]:8332: Connection refused (os error 111)
WARN - failed to connect daemon at [::1]:8332: Connection refused (os error 111)
.... keeps printing the Connection refused messages
*********************
- does the script need additional "ufw allow" commands?
- is it really optimized for the RaspiBlitz (just wondering because of the "--cookie=raspibolt:.." statement above
Thanks for your help!

**Comment by openoms** [comment]
Hi, thanks for contacting.
At which stage you are in the setup? 

I ran your command and it works on my node.

It seems that electrs is not able to connect to bitcoind.
Is it running? Check with `sudo systemctl status bitcoind`

Do you have server=1 in you bitcoin.conf?
Run `sudo nano /mnt/hdd/bitcoin/bitcoin.conf ` to check.

You don`t need to allow anything else than the port 50001 through ufw, because the rest of the communication is happening inside the node.

**Comment by Himbeergeld** [comment]
Yes, "bitcoind.service" reports active (running) since more than one week. The # Bitcoind options section has "server=1" set. The problem happens at the last step of the script:
cargo run --release -- -vvvv --index-batch-size=10 --jsonrpc-import --db-dir /mnt/hdd/electrs/db --cookie="raspibolt:$PASSWORD_B" --electrum-rpc-addr="0.0.0.0:50001"
I'd expect it to be the longest part of the script but the Connection refused error appears already shortly after this command.

**Comment by openoms** [comment]
I can`t reproduce your the error unfortunately. 

> Running target/release/electrs -vvvv --index-batch-size=10 --jsonrpc-import --db-dir /mnt/hdd/electrs/db '--cookie=raspibolt: \<myPsssowordhere\>' '--electrum-rpc-addr=0.0.0.0:50001'

in this one there a couple unnecessary `-s and the raspibolt: <myPsssowordhere> part needs to be between " ". 

Try running just these two lines (type your PASWORD_B followed by [ENTER] from the ~/electrs directory:

read PASSWORD_B
cargo run --release -- -vvvv --index-batch-size=10 --jsonrpc-import --db-dir /mnt/hdd/electrs/db --cookie="raspibolt:$PASSWORD_B" --electrum-rpc-addr="0.0.0.0:50001"

to make sure the format is correct.

**Comment by Himbeergeld** [comment]
Thank you, the additional localhost options allowed me to proceed. I also had to modify the path variable with `PATH=$PATH:$HOME/.cargo/bin:$HOME/electrs` before I ran the cargo command from the ~/electrs directory. Now it seems to proceed well and is currently indexing the database, I can see the Free Mem decreasing periodically down to red level on the RaspiBlitz screen, will probably take a while to finish. 
Just wondering why my RaspiBlitz doesn't have the default localhost associated. I just installed the V1.0 image, then updated to V1.1 and never touched the localhost configuration. Maybe these additional localhost parameters should be added by default to the script?

**Comment by openoms** [comment]
Ok, glad to hear that it is working now,  but it will still take some time to index. If it freezes a restart should not be a problem, it can continue indexing the blocks where it left off.
I don`t know why the localhost parameters did not appear on your setup, those are the dafult for electrs so should not be needed to be added on.

The script has the `source $HOME/.cargo/env` which runs: `export PATH="$HOME/.cargo/bin:$PATH"` and we are using the full path of for `cargo run` or running from the directory as you did, so updating the PATH is not necessary.

> I just installed the V1.0 image, then updated to V1.1

There must be something unique about your setup.
Did you not burn the SDcard fresh to update to v1.1?

**Comment by openoms** [comment]
Yes, the indexing is done and it is just keeping up to date with the blocks. It runs in the foreground this way, so if you close the terminal the service stops. 
You can exit now and just run electrs again when you need it. The indexing should take only a couple of minutes even after being off for days.

Alternatively continue with this step for automatic start and SSL connection:
https://github.com/openoms/bitcoin-tutorials/blob/master/electrs/README.md#set-up-the-electrs-service-autostart-and-the-connection-over-ssl

**Comment by openoms** [comment]
> I'm running RTL but I would assume this has no impact on localhost settings?!

No it should not. Did not see this problem before, will keep an eye. You have the extra options you need now, if it would reappear.

**Comment by Himbeergeld** [comment]
I probably underestimated the implications of the next steps. I don't have dynamic DNS and my RaspiBlitz is in the guest network of my router, where port-forwarding is not possible. Looks like I'm stuck at this point. Initially my plan was to run a Lightning wallet just locally, inside the guest network of my router. Do you think there is any way to achieve this without dynamic DNS or port forwarding? Happy for any advice.

**Comment by openoms** [comment]
So now you electrs is runni g and you should be able to access it within your local LAN with the Electrum wallet ,but not Eclair. 
Eclair is a self sufficient Lightning node which uses Electrum servers to access the Bitcoin nodes. You can use Eclair, bit not with your own Electrum serves if you don't want to do port forwarding.

You can go around port forwarding using Tor.
Take a look at this guide: https://github.com/openoms/bitcoin-tutorials/blob/master/Zeus_to_RaspiBlitz_through_Tor.md
The Zeus wallet can be a remote control for your Lightning node on the RaspiBlitz.

You will need a dynamic DNS for that

**Comment by openoms** [comment]
I wonder how are things working out for you @Himbeergeld ?

**Comment by Himbeergeld** [comment]
Thanks for asking, still in progress, so many bumps in the way of the Electrum installation on my Debian 9 (e.g. haven't even installed sudo, etc. ). As I'm a Linux beginner each issue takes some time for me to research. I will let you know when successful.

**Comment by Himbeergeld** [comment]
Currently the script complains that "Electrum requires Python version >= 3.6.1" even though it shows python3-pip version 9.0.1-2. But don't worry, I'll research myself, it's the best way of learning. :-)

**Comment by openoms** [comment]
:) that is about your Python3 version, not PIP. 
Update with: `sudo apt install python3`

To update every package:
`sudo apt update`
`sudo apt upgrade`

**Comment by Himbeergeld** [comment]
`sudo apt install python3` responds
python3 is already the newest version (3.5.3-1).
`sudo apt update` and `sudo apt upgrade` also doesn't change the situation.
Also tried following the procedure on Electrum website but at the end with your script or Electrum website procedure I always get this:

Processing ./Electrum-3.3.4.tar.gz
    Complete output from command python setup.py egg_info:
    Error: Electrum requires Python version >= 3.6.1...
        ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in ...

I admit this issue is drifting away from the original question.  Fine for me if you close it but if you have any idea, would be great.

**Comment by openoms** [comment]
your python version is clearly old.
if you run `sudo apt install python3`
after `sudo apt update` 
your system should be made aware of that.

if that not seems to help try:
`sudo apt install python3.6`

**Comment by Himbeergeld** [comment]
Didn't work, Python 3.6 doesn't seem to be in the Debian repository. Installed Python 3.7.2 from python.org and now my `python -V` responds "Python 3.7.2".
But still running your electrum_install_config_and_run.sh script has the same problem at the step `python3 -m pip install --user Electrum-3.3.4.tar.gz[fast]`:

Processing ./Electrum-3.3.4.tar.gz
    Complete output from command python setup.py egg_info:
    **Error: Electrum requires Python version >= 3.6.1...**
    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-ehnajkki-build/

It still detects a lower python version than I have actually installed.

**Comment by openoms** [comment]
right, you are almost there. If there are two verison of python `python` usually refers to python 2 and `python3` refers to python3.
But you can modify the script yourself.
Run:
`sudo nano electrs_install_on_RaspiBlitz.sh`
use the arrow keys to get to the python3 command and delete the `3` so it says only `python`.
press CTRL+O, ENTER, CRTL+X and run the script:
`./electrs_install_on_RaspiBlitz.sh`

**Comment by Himbeergeld** [comment]
Works and is proceeding now! But I get this warning which I'd like to resolve:

You are using pip version 18.1, however version 19.0.3 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.

But when I run `pip install --upgrade pip` I get "pip: command not found". 

So I did
`wget https://bootstrap.pypa.io/get-pip.py -O /tmp/get-pip.py`
...
 ‘/tmp/get-pip.py’ saved [1699325/1699325]
Then `sudo python3 /tmp/get-pip.py`
...
Installing collected packages: pip
  Found existing installation: pip 9.0.1
    Uninstalling pip-9.0.1:
      Successfully uninstalled pip-9.0.1
Successfully installed pip-19.0.3

Which looks like it has installed pip-19.0.3 but I still get the same warning in your script:

You are using pip version 18.1, however version 19.0.3 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.

**Comment by openoms** [comment]
you have installed the new pip for python3. 
Try `sudo python /tmp/get-pip.py`

This is quite unconvetional though.

Better is:
`python3 -m pip install --upgrade pip `
and
`python2 -m pip install --upgrade pip`

https://askubuntu.com/questions/712339/how-to-upgrade-pip-to-latest

**Comment by Himbeergeld** [comment]
Thanks! `sudo python /tmp/get-pip.py` helped.
But even though it previously reports "Successfully installed Electrum-3.3.4", it now it gets stuck at the next steps in the script:

./electrum_install_config_and_run.sh: line 43: electrum: command not found
./electrum_install_config_and_run.sh: line 44: electrum: command not found
./electrum_install_config_and_run.sh: line 45: electrum: command not found
./electrum_install_config_and_run.sh: line 46: electrum: command not found
./electrum_install_config_and_run.sh: line 48: electrum: command not found

**Comment by openoms** [comment]
I am using Ubuntu 18.04 which is Debain based, but there are differences.
You can do the settings manually:

 Make Electrum config persist (editing ~/.electrum/config):
`sudo nano ~/.electrum/config`
>
    "oneserver": true,
    "rpcpassword": "$PASSWORD_B",
    "rpcuser": "raspibolt",
    "server": "192.168.1.239:50001:t", 

or can run electrum from it\`s folder with:
`./electrum --oneserver --server $RASPIBLITZ_IP:50001:t `

**Comment by Himbeergeld** [comment]
Actually the remark "run electrum from it's folder" led to the solution. I had to locate electrum and change to folder ~/.local/bin to be able to run all the electrum commands from your script manually.

Then on the final electrum command of your script I still ran into

Error: Could not import PyQt5 on Linux systems, you may try 'sudo apt-get install python3-pyqt5'

but `sudo apt-get install python3-pyqt5` didn't help.
In folder /usr/local/bin I had to execute `sudo ./pip install PyQt5` which then "Successfully installed PyQt5-5.12.1 PyQt5-sip-4.19.15"

Then after going back to folder ~/.local/bin I was finally able to successfully start electrum with `./electrum --oneserver --server <my IP here>:50001:t` and the GUI opened! 

Thanks so much for all your help so far! Would love to send you some Lightning tipping, also for testing, please let me know your address. Today I received my Ledger Nano S. In the next step I'll try to set up the wallet for this HW keystore and I'll check the interaction with the RaspiBlitz. Fine for me if you want to close this issue, otherwise I'll post some additional infos on my way to the full integration.

**Comment by openoms** [comment]
I will close this as no longer relevant to the title, but very happy to answer/help with anything here or on https://twitter.com/openoms. 
My most stable node currently: https://1ml.com/node/028a2cb8d51e44d7d7e108c2e80a98cc069145e05a6d2025cf554bd8866fe32993
Let me know if you want to connect (I will need to add you as a peer if you are coming from clearnet).

**Comment by openoms** [comment]
> Thanks! `sudo python /tmp/get-pip.py` helped.
> But even though it previously reports "Successfully installed Electrum-3.3.4", it now it gets stuck at the next steps in the script:
> 
> ./electrum_install_config_and_run.sh: line 43: electrum: command not found
> ./electrum_install_config_and_run.sh: line 44: electrum: command not found
> ./electrum_install_config_and_run.sh: line 45: electrum: command not found
> ./electrum_install_config_and_run.sh: line 46: electrum: command not found
> ./electrum_install_config_and_run.sh: line 48: electrum: command not found

@Himbeergeld 
Got to rerun this and made some modifications.
Added the installation directory where Python puts electrum to  the PATH an now the commands run successfully.

`PATH=$PATH:~/.local/bin/`

This way the plain `electrum` command works

**Comment by Himbeergeld** [comment]
Great, thanks for the script improvement!
On my side in the meantime the Ledger Nano S is working well with Electrum on my Debian 9.8 but the Electrum wallet itself can not get a connection to any server, neither to my RaspiBlitz nor to any public server like e.g. btc.cihar.com. Have started Electrum without command line options to have the full server flexibility but none of the server settings work. Ping to RaspiBlitz and public server works well but nevertheless the network-LED in Electrum is always red and status is "Not connected". Just in case you have any idea how to track this down, would be great.

**Comment by openoms** [comment]
have a look what is in the config file:

sudo nano ~/.electrum/config

The correct settings:

**Comment by Himbeergeld** [comment]
All config file settings are fine. 
Also public Electrum servers don't work for me, but this could be due to the ongoing DDoS attack (https://twitter.com/ElectrumWallet/status/1116063328927985664). Don't know how to debug this issue. Any idea appreciated.

**Comment by openoms** [comment]
Is the electrs running on your RaspiBlitz? Needs to be started every time you want to use it an needs to show that the service is running on port 50001. Can just rerun the intall script for that. Or you can install it as a systemd service for autostart with the electrs_automation_for_Eclair script.

Check if port 50001 is allowed on your RaspiBlitz firewall:
`sudo ufw status`

You can try to access it form another computer or OS (eg windows) to see if it is an issue in your Debian.

**Comment by Himbeergeld** [comment]
Sorry for being so blind. Somehow I was assuming that electrs was running on the RaspiBlitz in background but of course it wasn't. After running the installation script again (with the `--daemon-rpc-addr="127.0.0.1:8332" --monitoring-addr="127.0.0.1:4224"` tweak in the beginning of this thread) now my Electrum wallet connection status is green and I can see RPC activity in the Terminal. 
So I basically ran into two problems: 1. public Electrum servers were not working due to the ongoing DDoS attack and 2. my own RaspiBlitz electrs was not running as a systemd service.
As a next step I'll figure out how to permanently run it as a systemd service. Assuming electrs is running permanently on the RaspiBlitz, I'm just wondering if all the electrs debug output (quite a lot) will be in conflict with regular SSH access to the RaspiBlitz menu?! Can all the electrs debug output be disabled?

**Comment by openoms** [comment]
I reopen this issue, because the input can be useful for others as well.

Absolutely the DDOS against Electrum servers is a major issue and awareness is raised on Twitter: https://twitter.com/ElectrumWallet/status/1116063328927985664.

The first install script does not set up the systemd service intentionally, so the RPi is not getting burdened by running it continuously. I am not sure yet if it takes it, but that is why we need to test.
You can start electrs again by running the script or just this command (with your PASSWORD_B):
`/home/admin/electrs/target/release/electrs --index-batch-size=10 --jsonrpc-import --db-dir /mnt/hdd/electrs/db  --electrum-rpc-addr="0.0.0.0:50001" --cookie="raspibolt:$PASSWORD_B"`
You will be still able to interactwith your RaspiBlitz if you open a second terminal window. You can have as many ssh connections open as you like.

The systemd service can be installed can be installed with the second script so electrs will run in the background always.

Can you tell roughly how long it took for the Pi to finish the indexing before starting to serve the Electrum wallet again? If it is an acceptable waiting time it would be make sense to just run the server intermittently when needed. I am worried about hogging the limited 1GB RAM of the Pi with the always-on service.

**Comment by Himbeergeld** [comment]
The re-sync didn't take too long, maybe 15 min. I didn't watch the RAM consumption but without electrs I have almost 500 MB RAM free. I will test this after Easter and watch the RAM consumption.

**Comment by openoms** [comment]
> The re-sync didn't take too long, maybe 15 min.

hmm ~15 mins after 6 days inactivity. That would make well over an hour after a month. Far from an ideal UX.

**Comment by Himbeergeld** [comment]
Did another test: After 5 days of inactivity, electrs had 777 block headers left to index which took 8 min. to finish. 
Indeed the RAM of the RaspiBlitz is heavily affected. During indexing, the free RAM went down to 43 MB. In static operation, with Electrum wallet running, the free RAM started in the red "red zone" (approx. 90 MB) but then after a few minutes moved up into the green zone and finally settled down at approx. 200 MB.
Not too bad I would say. The RaspiBlitz definitely seems to have some moments of stress in the beginning during sync but 10-15 min after the sync, it's all fine again.
I'd definitely consider the electrs a good option to be added to the RaspiBlitz "Additional Services" menu.

**Comment by openoms** [comment]
closing this now because since there is the new 1.3 version is out and Electrs has had updates too.
